// rethinkdb-ts/lib/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RethinkDBErrorType = void 0;
var RethinkDBErrorType;
(function (RethinkDBErrorType) {
    RethinkDBErrorType[RethinkDBErrorType["UNKNOWN"] = 0] = "UNKNOWN";
    // driver
    RethinkDBErrorType[RethinkDBErrorType["API_FAIL"] = 1] = "API_FAIL";
    // query errors
    RethinkDBErrorType[RethinkDBErrorType["CONNECTION"] = 2] = "CONNECTION";
    RethinkDBErrorType[RethinkDBErrorType["MASTER_POOL_FAIL"] = 3] = "MASTER_POOL_FAIL";
    RethinkDBErrorType[RethinkDBErrorType["POOL_FAIL"] = 4] = "POOL_FAIL";
    RethinkDBErrorType[RethinkDBErrorType["CURSOR_END"] = 5] = "CURSOR_END";
    RethinkDBErrorType[RethinkDBErrorType["TIMEOUT"] = 6] = "TIMEOUT";
    RethinkDBErrorType[RethinkDBErrorType["CANCEL"] = 7] = "CANCEL";
    RethinkDBErrorType[RethinkDBErrorType["PARSE"] = 8] = "PARSE";
    RethinkDBErrorType[RethinkDBErrorType["ARITY"] = 9] = "ARITY";
    RethinkDBErrorType[RethinkDBErrorType["CURSOR"] = 10] = "CURSOR";
    // connection error
    RethinkDBErrorType[RethinkDBErrorType["AUTH"] = 11] = "AUTH";
    RethinkDBErrorType[RethinkDBErrorType["UNSUPPORTED_PROTOCOL"] = 12] = "UNSUPPORTED_PROTOCOL";
    // reql response errors
    RethinkDBErrorType[RethinkDBErrorType["INTERNAL"] = 13] = "INTERNAL";
    RethinkDBErrorType[RethinkDBErrorType["RESOURCE_LIMIT"] = 14] = "RESOURCE_LIMIT";
    RethinkDBErrorType[RethinkDBErrorType["QUERY_LOGIC"] = 15] = "QUERY_LOGIC";
    RethinkDBErrorType[RethinkDBErrorType["NON_EXISTENCE"] = 16] = "NON_EXISTENCE";
    RethinkDBErrorType[RethinkDBErrorType["OP_FAILED"] = 17] = "OP_FAILED";
    RethinkDBErrorType[RethinkDBErrorType["OP_INDETERMINATE"] = 18] = "OP_INDETERMINATE";
    RethinkDBErrorType[RethinkDBErrorType["USER"] = 19] = "USER";
    RethinkDBErrorType[RethinkDBErrorType["PERMISSION_ERROR"] = 20] = "PERMISSION_ERROR";
})(RethinkDBErrorType = exports.RethinkDBErrorType || (exports.RethinkDBErrorType = {}));
// #endregion operations
